proposed/event-dispatcher.md

```diff
@@ -11,75 +11,77 @@ The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "S
+Implementers MUST assume that the same object will be passed to all Listeners.
 
-It is RECOMMENDED, but NOT REQUIRED, that Event objects support lossless serialization and deserialization.  That is, `$event == unserialize(serialize($event))` SHOULD hold true.  Objects MAY leverage PHP’s Serializable interface, `__sleep()` or `__wakeup()` magic methods, or similar language functionality if appropriate.
+It is RECOMMENDED, but NOT REQUIRED, that Event objects support lossless serialization and deserialization; `$event == unserialize(serialize($event))` SHOULD hold true.  Objects MAY leverage PHP’s `Serializable` interface, `__sleep()` or `__wakeup()` magic methods, or similar language functionality if appropriate.
 
 ## Stoppable Events
 
 A **Stoppable Event** is a special case of Event that contains additional ways to prevent further Listeners from being called.  It is indicated by implementing the `StoppableEventInterface`.
 
-An Event that implements `StoppableEventInterface` MUST return `true` from `isPropagationStopped()` when whatever Event it represents has been completed.  It is up to the implementer of the class to determine when that is.  For example, an Event that is asking for a PSR-7 `RequestInterface` object to be matched with a corresponding `ResponseInterface` object MAY have a `setResponse(ResponseInterface $res)` method for a Listener to call, which sets an internal flag that `isPropagationStopped()` will use to return `true` once the response has been set.
+An Event that implements `StoppableEventInterface` MUST return `true` from `isPropagationStopped()` when whatever Event it represents has been completed.  It is up to the implementer of the class to determine when that is.  For example, an Event that is asking for a PSR-7 `RequestInterface` object to be matched with a corresponding `ResponseInterface` object MAY have a `setResponse(ResponseInterface $res)` method for a Listener to call, which causes `isPropagationStopped()` to return `true`.
 
 ## Listeners
 
-A Listener may be any PHP callable.  A Listener MUST have one and only one parameter, which is the Event to which it responds.  Listeners SHOULD type hint that parameter as specifically as is relevant for their use case; that is, a Listener MAY type hint against an interface to indicate it is compatible with any Event type that implements that interface.
+A Listener may be any PHP callable.  A Listener MUST have one and only one parameter, which is the Event to which it responds.  Listeners SHOULD type hint that parameter as specifically as is relevant for their use case; that is, a Listener MAY type hint against an interface to indicate it is compatible with any Event type that implements that interface, or to a specific implementation of that interface.
 
-A Listener MUST have a `void` return, and SHOULD type hint that return explicitly.
+A Listener SHOULD have a `void` return, and SHOULD type hint that return explicitly.
```

Comment:
This is likely the most controversial change I've proposed.

The reason I've changed it from `MUST` to `SHOULD` is because many existing event systems _allow return values_. Some aggregate them, some do not; some ignore them entirely. But if we make this a `MUST`, it may potentially cause some readers to feel that their existing listeners are invalid, forcing them to choose between not implementing the spec, or breaking backwards compatibility.

Reply:
Hrm.  My concern here is a PSR-14 "compatible" implementation that does something with the return values.  Then someone relying on that is no longer PSR-14 compatible as soon as they switch to one that doesn't.

Reply:
:thinking: What if we changed it to a SHOULD, but also added a MUST NOT for the dispatcher, that it's not allowed to do anything meaningful with the return?  That would still allow the layering BC you're talking about but would, I think, preclude any incompatible variations.

Reply:
That would work; shall I propose a wording via a new commit?